modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+		__builtin_types_compatible_p(__typeof__(&location_request),
+					     __typeof__(&__real_location_request)),
+	"Error: Wrapped function does not match original location_request function signature");
+
+static void memfault_location_event_handler(const struct location_event_data *event_data)
+{
+	if (event_data->id == LOCATION_EVT_LOCATION || event_data->id == LOCATION_EVT_TIMEOUT ||
+	    event_data->id == LOCATION_EVT_ERROR || event_data->id == LOCATION_EVT_RESULT_UNKNOWN) {
+		memfault_gnss_metrics_stop_fix_session(event_data);
+	}
+	real_location_event_handler(event_data);
+}
+
+/**
+ * @brief Start a GNSS fix session.
+ *
+ */
+static void memfault_gnss_metrics_start_fix_session(void)
+{
+	if (session_in_progress) {
+		return;
+	}
+
+	session_in_progress = true;
+	LOG_DBG("Starting GNSS session");
+	MEMFAULT_METRICS_SESSION_START(ncs_gnss);
+	MEMFAULT_METRIC_ADD(ncs_gnss_fix_request_count, 1);
+}
+
+/**
+ * @brief Stop a GNSS fix session.
+ *
+ * @param event_data pointer to location event data
+ */
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data)
+{
+	if (!session_in_progress) {
+		return;
+	}
+
+	uint32_t session_time_ms;
+	memfault_metrics_heartbeat_timer_read(
+		MEMFAULT_METRICS_KEY(MEMFAULT_METRICS_SESSION_TIMER_NAME(ncs_gnss)),
+		&session_time_ms);
+
+	switch (event_data->id) {
+	case LOCATION_EVT_LOCATION:
+		LOG_DBG("Stopping GNSS session, fix data acquired");
+
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_time_to_fix_ms, session_time_ms);
+
+		uint32_t accuracy_cm = (uint32_t)(event_data->location.accuracy * 100.0f);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_fix_accuracy_cm, accuracy_cm);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_satellites_tracked_count,
+					     event_data->location.details.gnss.satellites_tracked);
+		break;
+	case LOCATION_EVT_TIMEOUT:
+		LOG_DBG("Stopping GNSS session, timeout recorded");
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_search_timeout_ms, session_time_ms);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_satellites_tracked_count,
+					     event_data->location.details.gnss.satellites_tracked);
+		break;
+	case LOCATION_EVT_RESULT_UNKNOWN:
+		// This location result will occur when a timeout occurs, but another method is
+		// attempted and fails (some "external" method via A-GNSS, P-GPS, LTE neighbor
+		// cell and Wi-Fi access point data). Record as a timeout as well but mark with
+		// a different log message for debugging purposes.
+		LOG_DBG("Stopping GNSS session, timeout recorded along with unknown result");
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_search_timeout_ms, session_time_ms);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_satellites_tracked_count,
+					     event_data->location.details.gnss.satellites_tracked);
+		break;
+	case LOCATION_EVT_ERROR:
+		LOG_DBG("Stopping GNSS session, error event occurred, id=%d", event_data->id);
+		break;
+	default:
+		LOG_DBG("Stopping GNSS session, unexpected event occured: id=%d", event_data->id);
+		break;
```

Comment:
It might be worth combining these two cases together, they both seem like errors. Might even be worth adding an error count metric?

Reply:
Good idea, will do